libraries/chain/include/eosio/chain/finality/finalizer.hpp

```diff
@@ -111,13 +108,25 @@ namespace eosio::chain {
          std::unique_lock g(mtx);
 
          // first accumulate all the votes
-         for (const auto& f : fin_pol.finalizers) {
-            if (auto it = finalizers.find(f.public_key); it != finalizers.end()) {
-               vote_message_ptr vote_msg = it->second.maybe_vote(it->first, bsp, bsp->strong_digest);
+         // optimized for finalizers of size one which should be the normal configuration outside of tests
+         for (auto& f : finalizers) {
+            auto it = std::ranges::find_if(bsp->active_finalizer_policy->finalizers, [&](const auto& fin) { return fin.public_key == f.first; });
+            if (it != bsp->active_finalizer_policy->finalizers.end()) {
+               vote_message_ptr vote_msg = f.second.maybe_vote(f.first, bsp, bsp->strong_digest);
                if (vote_msg)
                   votes.push_back(std::move(vote_msg));
+            } else if (bsp->pending_finalizer_policy) {
+               auto itr = std::ranges::find_if(bsp->pending_finalizer_policy->second->finalizers, [&](const auto& fin) { return fin.public_key == f.first; });
+               if (itr != bsp->pending_finalizer_policy->second->finalizers.end()) {
+                  if (std::ranges::find_if(votes, [&](const vote_message_ptr& vote_msg) { return vote_msg->finalizer_key == f.first; }) == votes.end()) {
+                     vote_message_ptr vote_msg = f.second.maybe_vote(f.first, bsp, bsp->strong_digest);
+                     if (vote_msg)
+                        votes.push_back(std::move(vote_msg));
+                  }
+               }
             }
          }
+
```

Comment:
I think we can make it simpler by voting iff the local finalizer is in active or pending policy.

```suggestion
         auto in_policy = [](const finalizer_policy_ptr& policy, const bls_public_key& key ){
            return std::ranges::any_of(policy->finalizers, [&](const auto& fin) { return fin.public_key == key; });
         };

         for (auto& f : finalizers) {
            if (in_policy(bsp->active_finalizer_policy, f.first) ||
                (bsp->pending_finalizer_policy && in_policy(bsp->pending_finalizer_policy->second, f.first))) {
               vote_message_ptr vote_msg = f.second.maybe_vote(f.first, bsp, bsp->strong_digest);
               if (vote_msg)
                  votes.push_back(std::move(vote_msg));
            }
         }
```

Reply:
Very nice.